### ISSUE TITLE
afterRemoveRow, afterInserRow, afterRemoveCol, afterInsertCol are called before cellMeta is updated

### ISSUE THREAD

**Issue by iaminvictus1993** [handsontable/handsontable#5525]
### Steps to reproduce

1. insert a row before row 1  col0
2.remove a row before row 1  col0
3.each steps check the result  getCellsMeta

### Demo

http://jsfiddle.net/s6t768pq/2868/

### Your environment
* Handsontable version: 6.1.0
* Browser Name and version:
* Operating System:

**Comment by AMBudnik** [comment]
Hi @rianran1993 

Do I open the menu on the 1st row (index 0) for point 2?

What is an expected result? (Just to confirm)

**Comment by rianran1993** [comment]
> Hi @rianran1993
> 
> Do I open the menu on the 1st row (index 0) for point 2?
> 
> What is an expected result? (Just to confirm)

yes , I found setTimeout cause diffrent result [demo](http://jsfiddle.net/s6t768pq/2871/)
CreateRow return the right cellsmeta length, but removerow didn't

**Comment by rianran1993** [comment]
> Wouldn't we need to add something in the example? The logs from both `row` hooks do not differ.
> 
> Do you run it on Chrome?

[care length](http://jsfiddle.net/s6t768pq/2874/)

**Comment by sforward** [comment]
I ran into the same issue. Within the `removeRow` function a wrong variable is used, I think it should be `priv.cellSettings.splice(calcIndex, groupAmount)`, to remove the cellSettings of the just removed rows.

**Comment by AMBudnik** [comment]
Thank you for doing the investigation @sforward 

You're right @rianran1993 

I am marking this topic as a bug.

**Comment by rianran1993** [comment]
> Thank you for doing the investigation @sforward
> 
> You're right @rianran1993
> 
> I am marking this topic as a bu

> Thank you for doing the investigation @sforward
> 
> You're right @rianran1993
> 
> I am marking this topic as a bug.

so I change as @sforward  say  can solve this?

**Comment by GrononS** [comment]
I am also experiencing this issue. This is a breaking bug.

**Comment by AMBudnik** [comment]
Replicable in `7.1.1` (6 calls) http://jsfiddle.net/handsoncode/6rfLp0bq/ and in `8.0.0.-rc1` http://jsfiddle.net/handsoncode/ougy4p3z/

**Comment by wojciechczerniak** [comment]
Not related to `v8.0.0`. This issue is a part of another Epic  #6274

**Comment by aldrinmartoq** [comment]
An inefficient but simple workaround: after removing rows, call `hot.loadData(your_data)`.

**Comment by aninde** [comment]
Works the same with a new version of cell meta [feature/issue-6233]( https://github.com/handsontable/handsontable/pull/6254). 
afterRemoveRow still returns getCellMeta like it was before removing, not after removing. 

Demo - `7.2.2-feature/issue-6233` http://jsfiddle.net/aninde/kr5d6ty2/ - no fixed
 `7.2.2-feature/issue-6233` http://jsfiddle.net/aninde/0qsy12hg/ - no fixed

**Comment by AMBudnik** [comment]
There's one thing that changes between those versions and IMO is a breaking change for the `getCellsMeta` method. 
The `getCellsMeta` returns different number of arguments.

![image](https://user-images.githubusercontent.com/10757813/70896797-af7cea00-1ff1-11ea-914a-fd98d286eb95.png)

simplified build demo http://jsfiddle.net/v2oq7ant/1/
simplified `7.3.0` demo https://jsfiddle.net/rbsyqx41/2/

**Comment by wojciechczerniak** [comment]
@AMBudnik please add your note to the PR, we may forgot that it's here

**Comment by aninde** [comment]
Looks like new `8.0.0-beta2-rev14` fixed described here issue https://jsfiddle.net/aninde/x7986ejk/ but I'm not sure for 100%, due to broken demo from original report. @AMBudnik can you double check?
<img width="206" alt="Screenshot 2020-06-16 at 16 58 26" src="https://user-images.githubusercontent.com/30186818/84791614-27ed5380-aff3-11ea-9f6a-b7ce6b5433e4.png">

**Comment by AMBudnik** [comment]
**rev14** returns a single array for the `getCellsMeta()` as it was for the previously mentioned build.

However, as I understand the subject @iaminvictus1993 and @sforward 
 wanted to get 4 elements not 6 in the

```js
afterRemoveRow: function(index) {
    if (init) {
      console.log(this.getCellsMeta())
    }
  }
```

call, when he removes the newly added row.

@iaminvictus1993 and @sforward Can you confirm?

**Comment by AMBudnik** [comment]
Closed with #6274

**Comment by wojciechczerniak** [comment]
@AMBudnik Is it? Those hooks are called inside `datamap` before or after the cell meta is updated. I don't think this has changed

**Comment by AMBudnik** [comment]
Hey @wojciechczerniak We weren't sure if the subject was understood well and there is no feedback. Thank you for checking. I'll reopen it for further investigation.

**Comment by maxxxxxdlp** [comment]
Is this fixed yet?
getCellsMeta throws errors for me when called from inside of afterChange hook when new rows were created

A simple demo:
http://jsfiddle.net/maxxxxxdlp/f3gze47m/2/

Steps to reproduce:
1. Select multiple rows (for example 3)
2. Copy them
3. Select any cell in the last row
4. Paste

getCellsMeta throws an error from inside of afterChange hook. Seems like it tries to get the meta for rows that were just newly created before their meta is initialized

Reproduced in 8.3.2 and 9.0.0

**Comment by AMBudnik** [comment]
Hi @maxxxxxdlp I do not get any error using the demo and provided scenario.
I'm testing it on Chrome 91 / Windows 10. I've also logged the result of `getCellsMeta` and it's correct.

**Comment by maxxxxxdlp** [comment]
@AMBudnik here is a video recording of this bug happening:

https://user-images.githubusercontent.com/40512816/121516440-e1101f00-c9f6-11eb-93f7-b77e6ddaf08f.mov



Altough, after I did some more investigation, it seems to happen even when getCellsMeta is called from outside of the afterChange hook, so this might be a different issue.

Demo:
https://jsfiddle.net/maxxxxxdlp/f3gze47m/12/

https://user-images.githubusercontent.com/40512816/121516381-cfc71280-c9f6-11eb-8545-02f8f147c467.mov

**Comment by AMBudnik** [comment]
Thank you for sharing @maxxxxxdlp 
I see that we get the same error here https://github.com/handsontable/handsontable/issues/7616 (with hidden data). Those cases seem to have in common the fact that we might be asking for something that is not accessible at the moment.

**Comment by budnix** [comment]
Hi, I tested this issue against the current `develop` branch and I cannot reproduce it anymore.

## What I tested

The problem reported here is that `afterCreateRow`, `afterRemoveRow`, `afterCreateCol` and
`afterRemoveCol` run *before* the cell meta store is updated. Reading cell meta inside those
callbacks then returns the state from before the change.

I checked the current code in `handsontable/src/dataMap/dataMap.ts`. All four paths now update
the meta store first and fire the hook second:

| Method | Order today |
|---|---|
| `createRow` | `metaManager.createRow(...)` → `runHooks('afterCreateRow')` |
| `createCol` | `metaManager.createColumn(...)` → `runHooks('afterCreateCol')` |
| `removeRow` | `metaManager.removeRow(...)` → `runHooks('afterRemoveRow')` |
| `removeCol` | meta removal in `#spliceRemovedColumns` → `runHooks('afterRemoveCol')` |

## How I verified it

I wrote five tests. Each one puts a marker on a cell, alters the grid, and asserts inside the
hook that the marker has already moved to its post-change position. All five pass.

To make sure the tests measure the right thing, I then put the old order back on purpose — I
moved the four `runHooks` calls ahead of the meta updates. All four marker tests turned red.
So the tests do detect the reported bug; they pass because the bug is gone.

The cell meta counts behave as originally requested too. Inside `afterRemoveRow` the count is
already reduced (14 → 11), and inside `afterRemoveCol` it is already reduced (13 → 12).

## When it was fixed

Commit 98bffa5aef (PR #10136, December 2022). That commit states the cause directly:

> The main thing that caused the bug was the fact that the `MetaManager` was called too late,
> after the afterCreate{Row,Col} hook was triggered.

It moved the meta update ahead of all four hooks, including the two remove hooks. The first
release that contains it is **12.3.0**.

## Two notes

The fix suggested earlier in this thread (`priv.cellSettings.splice`) was never applied. That
code no longer exists — the `MetaManager` replaced it. The ordering was fixed instead.

The last report in this thread, where `getCellsMeta` throws after pasting rows, is a separate
problem. It was already linked to #7616 and is not covered by what I tested here.

Since the reported behavior is resolved, I am closing this issue. If you still see it on
12.3.0 or later, please comment with a demo and I will reopen it.
